cdk/src/open-data-platform/data-plane/data-import/handler-factory.ts

```diff
@@ -48,59 +82,112 @@ export const geoJsonHandlerFactory = (
+              const rePromise = callback([row], db);
+              rePromise.catch((reason) => console.log(`batch${id}: Reprocessing failed`, reason));
+              promises.push(rePromise);
             }),
-          );
-          pipeline.resume();
+          ); // Supress unhandled rejections here.
+
+          try {
+            // Await this batches so we know whether to increment the row count.
+            // This does not block other batches from being processed in parallel.
+            await promise;
+            processedRowCount += rows.length;
+            console.log(`batch${id}: ${processedRowCount} rows have been processed so far.`);
+          } catch (error) {
+            // Handle in the promise.
+          }
+
+          // Stop reading stream if this would exceed number of rows to write.
+          // This check is done at the end because it doesn't know how many rows are currently
+          // being processed in parallel.
+          if (processedRowCount > rowLimit) {
+            console.log(`batch${id}: Stopping after passing row limit:`, rowLimit);
+            pipeline.destroy();
+            return;
+          }
         })
         .on('error', (error: Error) => {
           reject(error);
         })
         // Gets called by pipeline.destroy()
-        .on('close', async (_: Error) => {
-          resolve(numberRows);
-        })
-        .on('end', async () => {
-          resolve(numberRows);
-        });
+        .on('close', async (_: Error) => resolve(await handleEndOfFilestream(promises)))
+        .on('end', async () => resolve(await handleEndOfFilestream(promises)));
     });
 
+  /**
+   * Logs number and details of errors that occurred for all inserts.
+   * @param promises of SQL executions.
+   */
+  const handleEndOfFilestream = async (promises: Promise<any>[]): Promise<ImportResult> => {
+    // Unlike Promise.all(), which rejects when a single promise rejects, Promise.allSettled
+    // allows all promises to finish regardless of status and aggregated the results.
+    const result = await Promise.allSettled(promises);
+    const errors = result.filter((p) => p.status == 'rejected') as PromiseRejectedResult[];
+    console.log(`Number errors during insert: ${errors.length}`);
+
+    // Log all the rejected promises to diagnose issues.
+    errors.forEach((error) => console.log('Failed to process row:', error.reason));
+
+    return {
+      processedBatchCount: promises.length,
+      erroredBatchCount: errors.length,
+      sucessfulBatchCount: promises.length - errors.length,
+    };
+  };
+
   /**
    * Constructed handler that imports GeoJSON data to a DB.
    */
-  const handler = async (_: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
-    const db = await connectToDb();
-    if (db == undefined) {
-      throw Error('Unable to connect to db');
-    }
-    let numberRows = 0;
+  const handler = async (event: ImportRequest): Promise<APIGatewayProxyResult> => {
```

Comment:
we can cover this in standup so feel free to resolve after, but wondering how this gets invoked?

Reply:
`geoJsonHandlerFactory` returns a `handler` function, which `write-water-systems-data-handler.ts` uses as its handler, which is invoked like any other lambda handler.

I'll talk through it at standup.